osmomath/decimal.go

```diff
@@ -856,3 +866,56 @@ func DecApproxEq(t *testing.T, d1 BigDec, d2 BigDec, tol BigDec) (*testing.T, bo
 	diff := d1.Sub(d2).Abs()
 	return t, diff.LTE(tol), "expected |d1 - d2| <:\t%v\ngot |d1 - d2| = \t\t%v", tol.String(), diff.String()
 }
+
+// LogBase2 returns log_2 {x}.
+// Rounds down by truncations during division and right shifting.
+// Accurate up to 32 precision digits.
+// Implementation is based on:
+// https://stm32duinoforum.com/forum/dsp/BinaryLogarithm.pdf
+func (x BigDec) LogBase2() BigDec {
+	// create a new decimal to avoid mutating
+	// the receiver's int buffer.
+	xCopy := ZeroDec()
+	xCopy.i = new(big.Int).Set(x.i)
+	if xCopy.LTE(ZeroDec()) {
+		panic(fmt.Sprintf("log is not defined at <= 0, given (%s)", xCopy))
+	}
+
+	// Normalize x to be 1 <= x < 2
+
+	// y is the exponent that results in a whole multiple of 2.
+	y := ZeroDec()
+
+	// invariant: x >= 1
+	// while x < 1
+	for xCopy.LT(OneDec()) {
+		xCopy.i = xCopy.i.Lsh(xCopy.i, 1)
+		y = y.Sub(OneDec())
+	}
+
+	// invariant: x < 2
+	// while x >= 2
+	for xCopy.GTE(twoBigDec) {
+		xCopy.i = xCopy.i.Rsh(xCopy.i, 1)
+		y = y.Add(OneDec())
+	}
+
+	b := OneDec().Quo(twoBigDec)
+
+	// N.B. At this point x is a positive real number representing
+	// mantissa of the log. We estimate it using the following
+	// algorithm:
+	// https://stm32duinoforum.com/forum/dsp/BinaryLogarithm.pdf
+	// This has shown precision of 32 digits relative
+	// to Wolfram Alpha in tests
+	for i := 0; i < maxLog2Iterations; i++ {
+		xCopy = xCopy.Mul(xCopy)
+		if xCopy.GTE(twoBigDec) {
+			xCopy.i = xCopy.i.Rsh(xCopy.i, 1)
+			y = y.Add(b)
+		}
+		b.i = b.i.Rsh(b.i, 1)
```

Comment:
If I understand the source correctly, `b` is used as a breaking condition (terminate once enough mantisa bits are found). Here we're just running a fixed number of iterations. Should we short circuit once a certain number of these bits are found? I'm assuming we know in advance how many bits ensures the aprox converges to the log